### DISCUSSION TITLE
Web3 Install is taking forever

### DISCUSSION THREAD

**Discussion by cherb-programs** [smartcontractkit/full-blockchain-solidity-course-py#165] [Q&A] [answered]
Hi Guys, 

I'm having issues installing web3 with `pip install web3`. I've ran into the VS C++ tools error initially, but after installing everything  the VS builder tools that went away. Now when I run `pip install web3` It runs forever. It gives me this line constantly "INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime.", and "INFO: pip is looking at multiple versions of xxx to determine which version is compatible with other requirements. This could take a while.", where xxx is the package it is trying to install inside of the web3 package. 

I'm not having much luck with this one, I've tried setting up a virtual environment, adding this to to system variables "C:\Users\user\AppData\Local\Programs\Python\Python310\Scripts". I've also tried different terminals such as git bash and powershell (I'm on windows 10). I'm not sure where to go from here.

**Comment by ace-brown** [comment, marked as answer]
Hi. I faced the same error. what you should do is this:

1-instead of python 3.10, you must install python 3.9 or maybe a lower version  (3.9 worked for me)
2-remember to add the python, npm, pip and web3 path to system variables if you encountered the not recognized error.
3-It is caused by the latest version of pip . A temporary fix is to downgrade your pip version to 20.2:
pip install --upgrade pip==20.2

let me know if it fixed the problem, because it worked for me pretty good.

**Reply by PatrickAlphaC** [reply]
boom! Where did you learn this?

**Reply by ace-brown** [reply]
I learnt some part of it by searching the net and some other by my creativity :). However, the searching part was frustrating.

**Reply by ArcusConsulting** [reply]
@KAD105 @PatrickAlphaC I uninstalled Python 3.10, installed Python 3.9.6 and added it to my PATH, downgraded pip to 20.2, and I get the following error when I type 'pip install web3' in CMD:

C:\Users\Eric>pip install web3
Collecting web3
  Using cached web3-5.24.0-py3-none-any.whl (487 kB)
Requirement already satisfied: eth-hash[pycryptodome]<1.0.0,>=0.2.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from web3) (0.3.2)
Collecting eth-account<0.6.0,>=0.5.6
  Using cached eth_account-0.5.6-py3-none-any.whl (101 kB)
Collecting eth-abi<3.0.0,>=2.0.0b6
  Using cached eth_abi-2.1.1-py3-none-any.whl (27 kB)
Requirement already satisfied: eth-typing<3.0.0,>=2.0.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from web3) (2.2.2)
Collecting lru-dict<2.0.0,>=1.1.6
  Using cached lru-dict-1.1.7.tar.gz (10 kB)
Collecting ipfshttpclient==0.8.0a2
  Using cached ipfshttpclient-0.8.0a2-py3-none-any.whl (82 kB)
Requirement already satisfied: jsonschema<4.0.0,>=3.2.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from web3) (3.2.0)
Requirement already satisfied: protobuf<4,>=3.10.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from web3) (3.18.1)
Collecting aiohttp<4,>=3.7.4.post0
  Using cached aiohttp-3.7.4.post0-cp39-cp39-win_amd64.whl (634 kB)
Requirement already satisfied: eth-utils<2.0.0,>=1.9.5 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from web3) (1.10.0)
Collecting pywin32>=223; platform_system == "Windows"
  Using cached pywin32-302-cp39-cp39-win_amd64.whl (9.3 MB)
Requirement already satisfied: requests<3.0.0,>=2.16.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from web3) (2.26.0)
Collecting hexbytes<1.0.0,>=0.1.0
  Using cached hexbytes-0.2.2-py3-none-any.whl (6.1 kB)
Collecting websockets<10,>=9.1
  Using cached websockets-9.1-cp39-cp39-win_amd64.whl (90 kB)
Requirement already satisfied: pycryptodome<4,>=3.6.6; extra == "pycryptodome" in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from eth-hash[pycryptodome]<1.0.0,>=0.2.0->web3) (3.11.0)
Collecting bitarray<1.3.0,>=1.2.1
  Using cached bitarray-1.2.2.tar.gz (48 kB)
Collecting eth-keys!=0.3.2,<0.4.0,>=0.2.1
  Using cached eth_keys-0.3.3-py3-none-any.whl (20 kB)
Collecting rlp<3,>=1.0.0
  Using cached rlp-2.0.1-py2.py3-none-any.whl (20 kB)
Collecting eth-rlp<2,>=0.1.2
  Using cached eth_rlp-0.2.1-py3-none-any.whl (5.0 kB)
Collecting eth-keyfile<0.6.0,>=0.5.0
  Using cached eth_keyfile-0.5.1-py3-none-any.whl (8.3 kB)
Collecting parsimonious<0.9.0,>=0.8.0
  Using cached parsimonious-0.8.1.tar.gz (45 kB)
Collecting multiaddr>=0.0.7
  Using cached multiaddr-0.0.9-py2.py3-none-any.whl (16 kB)
Requirement already satisfied: six>=1.11.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from jsonschema<4.0.0,>=3.2.0->web3) (1.16.0)
Requirement already satisfied: pyrsistent>=0.14.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from jsonschema<4.0.0,>=3.2.0->web3) (0.18.0)
Requirement already satisfied: setuptools in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from jsonschema<4.0.0,>=3.2.0->web3) (56.0.0)
Requirement already satisfied: attrs>=17.4.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from jsonschema<4.0.0,>=3.2.0->web3) (21.2.0)
Collecting async-timeout<4.0,>=3.0
  Using cached async_timeout-3.0.1-py3-none-any.whl (8.2 kB)
Collecting yarl<2.0,>=1.0
  Using cached yarl-1.7.0-cp39-cp39-win_amd64.whl (122 kB)
Collecting typing-extensions>=3.6.5
  Using cached typing_extensions-3.10.0.2-py3-none-any.whl (26 kB)
Collecting multidict<7.0,>=4.5
  Using cached multidict-5.2.0-cp39-cp39-win_amd64.whl (45 kB)
Collecting chardet<5.0,>=2.0
  Using cached chardet-4.0.0-py2.py3-none-any.whl (178 kB)
Requirement already satisfied: cytoolz<1.0.0,>=0.10.1; implementation_name == "cpython" in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from eth-utils<2.0.0,>=1.9.5->web3) (0.11.0)
Requirement already satisfied: certifi>=2017.4.17 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (2021.10.8)
Requirement already satisfied: idna<4,>=2.5; python_version >= "3" in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (3.3)
Requirement already satisfied: charset-normalizer~=2.0.0; python_version >= "3" in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (2.0.7)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (1.26.7)
Collecting netaddr
  Using cached netaddr-0.8.0-py2.py3-none-any.whl (1.9 MB)
Collecting varint
  Using cached varint-1.0.2.tar.gz (1.9 kB)
Collecting base58
  Using cached base58-2.1.0-py3-none-any.whl (5.6 kB)
Requirement already satisfied: toolz>=0.8.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from cytoolz<1.0.0,>=0.10.1; implementation_name == "cpython"->eth-utils<2.0.0,>=1.9.5->web3) (0.11.1)
Using legacy 'setup.py install' for lru-dict, since package 'wheel' is not installed.
Using legacy 'setup.py install' for bitarray, since package 'wheel' is not installed.
Using legacy 'setup.py install' for parsimonious, since package 'wheel' is not installed.
Using legacy 'setup.py install' for varint, since package 'wheel' is not installed.
Installing collected packages: bitarray, eth-keys, parsimonious, eth-abi, rlp, hexbytes, eth-rlp, eth-keyfile, eth-account, lru-dict, netaddr, varint, base58, multiaddr, ipfshttpclient, async-timeout, multidict, yarl, typing-extensions, chardet, aiohttp, pywin32, websockets, web3
    Running setup.py install for bitarray ... error
    ERROR: Command errored out with exit status 1:
     command: 'c:\users\eric\appdata\local\programs\python\python39\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Eric\\AppData\\Local\\Temp\\pip-install-r2_tpiyw\\bitarray\\setup.py'"'"'; __file__='"'"'C:\\Users\\Eric\\AppData\\Local\\Temp\\pip-install-r2_tpiyw\\bitarray\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Eric\AppData\Local\Temp\pip-record-eqba3y59\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\users\eric\appdata\local\programs\python\python39\Include\bitarray'
         cwd: C:\Users\Eric\AppData\Local\Temp\pip-install-r2_tpiyw\bitarray\
    Complete output (13 lines):
    running install
    running build
    running build_py
    creating build
    creating build\lib.win-amd64-3.9
    creating build\lib.win-amd64-3.9\bitarray
    copying bitarray\test_bitarray.py -> build\lib.win-amd64-3.9\bitarray
    copying bitarray\test_util.py -> build\lib.win-amd64-3.9\bitarray
    copying bitarray\util.py -> build\lib.win-amd64-3.9\bitarray
    copying bitarray\__init__.py -> build\lib.win-amd64-3.9\bitarray
    running build_ext
    building 'bitarray._bitarray' extension
    error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'c:\users\eric\appdata\local\programs\python\python39\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Eric\\AppData\\Local\\Temp\\pip-install-r2_tpiyw\\bitarray\\setup.py'"'"'; __file__='"'"'C:\\Users\\Eric\\AppData\\Local\\Temp\\pip-install-r2_tpiyw\\bitarray\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Eric\AppData\Local\Temp\pip-record-eqba3y59\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\users\eric\appdata\local\programs\python\python39\Include\bitarray' Check the logs for full command output.
WARNING: You are using pip version 20.2; however, version 21.3 is available.
You should consider upgrading via the 'c:\users\eric\appdata\local\programs\python\python39\python.exe -m pip install --upgrade pip' command.

C:\Users\Eric>

Different error than yesterday, however still an error. It's still telling me I'm missing Microsoft C++ 14.0, which I installed yesterday. Not sure what else to do here... I really want to complete this 16h video, but this is a major roadblock.

**Reply by ArcusConsulting** [reply]
Even tried with Python 3.9.0... Slightly different error

Microsoft Windows [Version 10.0.19042.1237]
(c) Microsoft Corporation. All rights reserved.

C:\Users\Eric>pip install web3
Collecting web3
  Using cached web3-5.24.0-py3-none-any.whl (487 kB)
Collecting protobuf<4,>=3.10.0
  Using cached protobuf-3.18.1-cp39-cp39-win_amd64.whl (912 kB)
Collecting pywin32>=223; platform_system == "Windows"
  Using cached pywin32-302-cp39-cp39-win_amd64.whl (9.3 MB)
Collecting eth-abi<3.0.0,>=2.0.0b6
  Using cached eth_abi-2.1.1-py3-none-any.whl (27 kB)
Collecting ipfshttpclient==0.8.0a2
  Using cached ipfshttpclient-0.8.0a2-py3-none-any.whl (82 kB)
Collecting lru-dict<2.0.0,>=1.1.6
  Using cached lru-dict-1.1.7.tar.gz (10 kB)
Requirement already satisfied: requests<3.0.0,>=2.16.0 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from web3) (2.26.0)
Collecting eth-typing<3.0.0,>=2.0.0
  Using cached eth_typing-2.2.2-py3-none-any.whl (6.2 kB)
Collecting eth-account<0.6.0,>=0.5.6
  Using cached eth_account-0.5.6-py3-none-any.whl (101 kB)
Collecting jsonschema<4.0.0,>=3.2.0
  Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Collecting websockets<10,>=9.1
  Using cached websockets-9.1-cp39-cp39-win_amd64.whl (90 kB)
Collecting eth-hash[pycryptodome]<1.0.0,>=0.2.0
  Using cached eth_hash-0.3.2-py3-none-any.whl (8.8 kB)
Collecting aiohttp<4,>=3.7.4.post0
  Using cached aiohttp-3.7.4.post0-cp39-cp39-win_amd64.whl (634 kB)
Collecting eth-utils<2.0.0,>=1.9.5
  Using cached eth_utils-1.10.0-py3-none-any.whl (24 kB)
Collecting hexbytes<1.0.0,>=0.1.0
  Using cached hexbytes-0.2.2-py3-none-any.whl (6.1 kB)
Collecting parsimonious<0.9.0,>=0.8.0
  Using cached parsimonious-0.8.1.tar.gz (45 kB)
Collecting multiaddr>=0.0.7
  Using cached multiaddr-0.0.9-py2.py3-none-any.whl (16 kB)
Requirement already satisfied: charset-normalizer~=2.0.0; python_version >= "3" in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (2.0.7)
Requirement already satisfied: certifi>=2017.4.17 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (2021.10.8)
Requirement already satisfied: idna<4,>=2.5; python_version >= "3" in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (3.3)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from requests<3.0.0,>=2.16.0->web3) (1.26.7)
Collecting rlp<3,>=1.0.0
  Using cached rlp-2.0.1-py2.py3-none-any.whl (20 kB)
Collecting eth-rlp<2,>=0.1.2
  Using cached eth_rlp-0.2.1-py3-none-any.whl (5.0 kB)
Collecting eth-keyfile<0.6.0,>=0.5.0
  Using cached eth_keyfile-0.5.1-py3-none-any.whl (8.3 kB)
Collecting eth-keys!=0.3.2,<0.4.0,>=0.2.1
  Using cached eth_keys-0.3.3-py3-none-any.whl (20 kB)
Collecting bitarray<1.3.0,>=1.2.1
  Using cached bitarray-1.2.2.tar.gz (48 kB)
Collecting pyrsistent>=0.14.0
  Using cached pyrsistent-0.18.0-cp39-cp39-win_amd64.whl (62 kB)
Requirement already satisfied: setuptools in c:\users\eric\appdata\local\programs\python\python39\lib\site-packages (from jsonschema<4.0.0,>=3.2.0->web3) (49.2.1)
Collecting attrs>=17.4.0
  Using cached attrs-21.2.0-py2.py3-none-any.whl (53 kB)
Collecting six>=1.11.0
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting pycryptodome<4,>=3.6.6; extra == "pycryptodome"
  Using cached pycryptodome-3.11.0-cp35-abi3-win_amd64.whl (1.8 MB)
Collecting multidict<7.0,>=4.5
  Using cached multidict-5.2.0-cp39-cp39-win_amd64.whl (45 kB)
Collecting yarl<2.0,>=1.0
  Using cached yarl-1.7.0-cp39-cp39-win_amd64.whl (122 kB)
Collecting typing-extensions>=3.6.5
  Using cached typing_extensions-3.10.0.2-py3-none-any.whl (26 kB)
Collecting chardet<5.0,>=2.0
  Using cached chardet-4.0.0-py2.py3-none-any.whl (178 kB)
Collecting async-timeout<4.0,>=3.0
  Using cached async_timeout-3.0.1-py3-none-any.whl (8.2 kB)
Collecting cytoolz<1.0.0,>=0.10.1; implementation_name == "cpython"
  Using cached cytoolz-0.11.0-cp39-cp39-win_amd64.whl (367 kB)
Collecting varint
  Using cached varint-1.0.2.tar.gz (1.9 kB)
Collecting netaddr
  Using cached netaddr-0.8.0-py2.py3-none-any.whl (1.9 MB)
Collecting base58
  Using cached base58-2.1.0-py3-none-any.whl (5.6 kB)
Collecting toolz>=0.8.0
  Using cached toolz-0.11.1-py3-none-any.whl (55 kB)
Using legacy 'setup.py install' for lru-dict, since package 'wheel' is not installed.
Using legacy 'setup.py install' for parsimonious, since package 'wheel' is not installed.
Using legacy 'setup.py install' for bitarray, since package 'wheel' is not installed.
Using legacy 'setup.py install' for varint, since package 'wheel' is not installed.
Installing collected packages: protobuf, pywin32, pycryptodome, eth-hash, toolz, cytoolz, eth-typing, eth-utils, six, parsimonious, eth-abi, varint, netaddr, base58, multiaddr, ipfshttpclient, lru-dict, rlp, hexbytes, eth-rlp, eth-keys, eth-keyfile, bitarray, eth-account, pyrsistent, attrs, jsonschema, websockets, multidict, yarl, typing-extensions, chardet, async-timeout, aiohttp, web3
    Running setup.py install for parsimonious ... done
    Running setup.py install for varint ... done
    Running setup.py install for lru-dict ... error
    ERROR: Command errored out with exit status 1:
     command: 'c:\users\eric\appdata\local\programs\python\python39\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Eric\\AppData\\Local\\Temp\\pip-install-u1xmacsw\\lru-dict\\setup.py'"'"'; __file__='"'"'C:\\Users\\Eric\\AppData\\Local\\Temp\\pip-install-u1xmacsw\\lru-dict\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Eric\AppData\Local\Temp\pip-record-n3kx1px9\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\users\eric\appdata\local\programs\python\python39\Include\lru-dict'
         cwd: C:\Users\Eric\AppData\Local\Temp\pip-install-u1xmacsw\lru-dict\
    Complete output (5 lines):
    running install
    running build
    running build_ext
    building 'lru' extension
    error: Microsoft Visual C++ 14.0 is required. Get it with "Build Tools for Visual Studio": https://visualstudio.microsoft.com/downloads/
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'c:\users\eric\appdata\local\programs\python\python39\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Eric\\AppData\\Local\\Temp\\pip-install-u1xmacsw\\lru-dict\\setup.py'"'"'; __file__='"'"'C:\\Users\\Eric\\AppData\\Local\\Temp\\pip-install-u1xmacsw\\lru-dict\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\Eric\AppData\Local\Temp\pip-record-n3kx1px9\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\users\eric\appdata\local\programs\python\python39\Include\lru-dict' Check the logs for full command output.
WARNING: You are using pip version 20.2; however, version 21.3 is available.
You should consider upgrading via the 'c:\users\eric\appdata\local\programs\python\python39\python.exe -m pip install --upgrade pip' command.

C:\Users\Eric>

**Reply by ArcusConsulting** [reply]
Since it's not a file I can't add it... It just shows up in Apps & Features, but if I search "Microsoft Visual C++" in explorer nothing comes up.

**Reply by ace-brown** [reply]
As stated in the error you need to download and install Build Tools for Visual Studio. It's here: https://visualstudio.microsoft.com/downloads/
in All downloads section at the bottom of the page.
after downloading:
Run the installer --> Select: Workloads → Visual C++ build tools.
Install options: select only the “Windows 10 SDK” (assuming the computer is Windows 10)

**Reply by ArcusConsulting** [reply]
I get the following error trying to install it:

"A setup package is either missing or damaged" and no matter if I select "Download packages from the internet" or "Please provide a location to search for the packages" nothing works. It doesn't even tell me which setup package is damaged and/or missing.

**Reply by ArcusConsulting** [reply]
Downloaded "Windows 10 SDK version 2104 (10.0.20348.0)"  I found on https://developer.microsoft.com/en-ca/windows/downloads/sdk-archive/ and getting the exact same error. Trying to keep calm but this getting bats**t frustrating. I've tried every tip and trick I've come across here and stackoverflow and nothing works.

Even https://visualstudio.microsoft.com/downloads/ doesn't work... if you scroll down to "Build Tools for Visual Studio 2019" and click the purple download button, it brings you to a new page that says  "Thank you for downloading Visual Studio" without downloading anything at all.

**Reply by ArcusConsulting** [reply]
There is no actual download file... the URL preview is a separate webpage, and on that webpage there is still no download file...

**Reply by ArcusConsulting** [reply]
@PatrickAlphaC Terribly sorry Patrick, but it doesn't look like I'll be completing your course anytime soon. Disappointing, really. I've tried everything I've come across and I keep getting that same stupid error. Tried with Python 3.10, 3.9.6 and now 3.9.0. I've installed Microsoft Visual C++ 14.0, and it will not let me install Windows SDK. I don't mind troubleshooting and doing some research, but this is just stupid. I'm doing everything I should be doing - correctly - and it is not cooperating. There is no reason it should be this complicated to install 1 single package.

**Reply by ace-brown** [reply]
Yeah you're right. sometimes errors becomes terribly annoying. But be patient my friend. Maybe other more experienced members can help. Don't be disappointed. You'll be getting what you're looking for in due time. I personally in these kind of situation try to follow along with the tutorials without coding and just watching. Maybe in another project you can tackle the error. But never let an error stop you from achieving your goal.

**Reply by ArcusConsulting** [reply]
@KAD105 I equate this to turning the key in your and it won't start, and the mechanic says "I've done everything to make it work, sorry." Maybe if more and more people run into the same issue other solutions will come about. I'm at the point where I'm encountering errors installing other packages outside of Python... I can only go so far down the rabbit hole!

**Reply by ArcusConsulting** [reply]
@KAD105 Ok so I managed to install web3 on my Raspberry Pi 4. Just copied my code over and will test tomorrow. I was able to install solcx but tried updating the version to 0.6.0 but got an error message. Will update tomorrow.

**Reply by Evancitoo** [reply]
i can not down grade the pip =(..!! i fells so confusing to my
![Screenshot_2](https://user-images.githubusercontent.com/67725330/137574717-2f4cfc4a-356c-4858-a2c0-47dc50c4e49c.jpg)
maybe i have downgrade from same path i had to upgrade it.

**Reply by Evancitoo** [reply]
it doesn't feel god downgrade the update fells counter intuitive.  because i don understand at all i had to prove many combinatoric of stuff and try to understand some out put message
![Screenshot_3](https://user-images.githubusercontent.com/67725330/137574914-e9e50a10-839d-4609-bfdf-b06a13373df6.jpg)
..

**Reply by Evancitoo** [reply]
now there should be a similar command to lower the python and dont make me uninstall and re install and reboot vs code. and wait until some error  that not working any more.

**Reply by Evancitoo** [reply]
see..!! it tells me that i have to upgrade my pip again so i had to upgrade. but i tried upgrade but then web3 can not update or install correctly then
![Screenshot_1](https://user-images.githubusercontent.com/67725330/137575290-6069a2ab-1ea8-4bf7-a004-cdf27e30208d.jpg)

**Comment by cherb-programs** [comment]
Hi guys, checking in, I've managed to get past this stage which I believe was a combination of a few errors in my environment. I still believe my environment isn't 100% right because I keep running into these weird package errors as I progress through the course (PIPX for example). Where it doesn't recognize the command even though it's been installed. Usually it's because it has installed itself somewhere weird and it's not on path (I installed like 4 different python versions trying to solve this and they're all in different places, so that will probably bite me in the ass down the track, but hey, it works!...for now). So I'll elaborate on what I did and hopefully this helps someone. I spent hours trying to figure this out...very frustrating indeed. 

Firstly, KAD105's steps are definitely things to try if you run into similar errors. 
Build tools are necessary so be sure to have them installed. I got mine from here https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2019 then scrolled down and selected "Build Tools for Visual Studio 2019". I saw @ArcusConsulting was having problems with this link but it seemed to work for me. Maybe try using another browser, Chrome worked for me. 

Once that's downloaded, restart your PC before continuing! Now,  From what I read Python 3.10 and Pip weren't working correctly so definitely try using Python 3.9 or something lower and don't forget to change your interpreter to whatever Python path your using!!!

Now, I played around with different virtual environments, Anaconda and different shells. I used a combination of WSL, git bash and Powershell to get all the packages installed but I think if I were to do it again I would do the following. 

Install VS build tools.

Install Python 3.9 or lower and ADD to PATH!! 
You can check what's in your path in PowerShell but typing `$env:Path`. Ensure you've added "path/to/python/Python.exe ", "path/to/python/Python39/site-packages" and "path/to/python/Python39/Scripts" in to your System Variables (I put them in User Variables as well just to be safe lol).  Also, close every terminal that you have open including the ones in VS Code **_before_** restarting VS Code. I had an instance where I ran a PowerShell outside of VS Code and it was recognizing that was recognizing PIPX and the one inside VS code wasn't. I didn't release I had 2 VS code windows open so when I restarted the one I was working on it still wasn't picking it up. It wasn't until I restarted BOTH of the VS Code Windows that it worked. 

Run everything in Admin mode.

You may have to play around with py, python or python3 syntax to get a certain packages to work. Or install it through npm or another method if pip/pipx isn't playing ball. From there, make sure said package is added to PATH using the method above. 

Anyway, hope that helps my fellow Windows users. Once I finish this course I will probably just make a dual boot system for Linux and just use that moving forward, does that seem like a good idea?

**Reply by ArcusConsulting** [reply]
@cherb-programs YOU sir... I owe you a beer. Turns out all I was missing was that Build Tools download. For some reason it wouldn't open with Chrome but it opened with Edge. Now I have a working environment on my Windows machine as well as my Raspberry Pi 4! Thank you thank you thank you! I can finally resume the course!!

**Reply by cherb-programs** [reply]
No problems, I've been caught out with that in the past too! Glad it worked for you.

**Reply by ace-brown** [reply]
Awesome 👍

**Reply by PatrickAlphaC** [reply]
Teamwork makes the dreamwork. Thanks all

**Reply by Evancitoo** [reply]
i already had visual studio 2019 community version.  but i had un install python3.8 to install 3.10 but vs code can not recognize the new version útil i had un install 3.8. and restar vs code and putted fix install3.10 then i could install pip solcx  but it needed a upgrade then i had upgrade the upgrade didnt worked until had to put it on cmd in a very specific pad  C:\Users\luis4\AppData\Local\Programs\Python\Python310>phython.exe  -m pip install --upgrade,  then the next problem was web3.  now i had a new version of pip that work with 3.10 maybe the old solution is no longer available not sure. but i already had visual-studio 2019 community version what if i install 2022 community one.? now the link show 2022 =(..!!

**Reply by Evancitoo** [reply]
now the pip version is 21.3 thing may had change not sure how.  but solcx was not hable to fix on import as my pip was not upgrade.   but web 3 still not fix i will try to download vs studio community 2022 i don understand very well how does it influence hope it may help.
![Screenshot_1](https://user-images.githubusercontent.com/67725330/137574593-0f71efd3-d699-4df4-a5a4-a0c85722a999.jpg)
.

**Comment by Evancitoo** [comment]
![Screenshot_1](https://user-images.githubusercontent.com/67725330/137573845-4bae1323-3d75-4893-b0e8-ab3394655cdd.jpg)
![Screenshot_3](https://user-images.githubusercontent.com/67725330/137573846-65ab8fa5-2a8a-487d-a473-b682c303658d.jpg)
![Screenshot_5](https://user-images.githubusercontent.com/67725330/137573847-1432d0ab-6f25-4d2a-8cf5-14c46497c8c2.jpg
![Screenshot_2](https://user-images.githubusercontent.com/67725330/137573850-4dbcc95b-bd38-4405-9332-98324bc9d61f.jpg)
![Screenshot_4](https://user-images.githubusercontent.com/67725330/137573851-2b878563-6919-4259-9696-499a47ba39bf.jpg)
)  oh i m new i got a error too, on installing web3 is this the same kind of error or its another one? i will try to find out. and share my experience. how can i keep the new version of python and also the old version? how do i switch back and forth the version of python 3.10 to 3.9? on vs code? is it posible?
